src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,443 @@
+        3. the array size is NOT a multiple of window_size + distance and
+           it has length n * (window_size + distance) + m, where m >= window_size
+
+        Args:
+            array: The numpy array to reshape.
+            window_size: The size of the window in samples.
+
+        Returns:
+            The reshaped 2D array.
+
+        Raises:
+            ValueError: If the array is smaller or equal to the given distance.
+        """
+        # Not using the num_windows function here because we want to
+        # differentiate between the three cases.
+        if len(array) < self._distance:
+            raise ValueError(
+                f"The array (length:{len(array)}) is too small to be reshaped."
+            )
+
+        num_windows = math.floor(len(array) / self._distance)
+
+        # Case 1:
+        if len(array) - num_windows * self._distance == 0:
+            log.debug("Case 1:")
+            resized_array = array
+        # Case 2
+        elif len(array) - num_windows * self._distance < window_size:
+            log.debug("Case 2:")
+            resized_array = array[: num_windows * self._distance]
+        # Case 3
+        else:
+            log.debug("Case 3:")
+            num_windows += 1
+            resized_array = np.resize(array, num_windows * self._distance)
+
+        return resized_array.reshape(num_windows, self._distance)
+
+    def _calculate_avg_window_py(
+        self,
+        window: NDArray[np.float_],
+        window_size: int,
+        weights: List[float] | None = None,
+    ) -> NDArray[np.float_]:
+        """
+        Plain python version of the average calculator.
+
+        This method avoids copying in any case but is 15 to 600
+        slower then the numpy version.
+
+        This method is only used in benchmarks and unit tests.
+
+        ONLY USE THIS IF YOU NEED TO STRICTLY AVOID COPIES!
+
+        Args:
+            window: The window to calculate the average over.
+            window_size: The size of the window to calculate the average over.
+            weights: The weights to use for the average calculation.
+
+        Returns:
+            The averaged window.
+        """
+        num_windows = self._num_windows(window, window_size)
+
+        if len(window) - num_windows * self._distance >= window_size:
+            num_windows += 1
+
+        res = np.empty(window_size)
+
+        for i in range(window_size):
+            assert num_windows * (self._distance) - len(window) <= self._distance
+            summe = 0
+            for j in range(num_windows):
+                if weights is None:
+                    summe += window[i + (j * (self._distance))]
+                else:
+                    summe += weights[j] * window[i + (j * (self._distance))]
+
+            if weights is None:
+                res[i] = summe / num_windows
+            else:
+                res[i] = summe / np.sum(weights)
```

Comment:
You also need to check `weights` is an empty list, before sending to `np.sum`,  you can change the check such that it would pass if weights is None, an empty list or set etc.

``` suggestion
            if not weights:
                res[i] = summe / num_windows
            else:
                res[i] = summe / np.sum(weights)